// Fix redefinition from ndk r25 clang 

diff --git a/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c b/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c
index be578a0bb..ce5e8b400 100644
--- a/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c
+++ b/src/java.base/unix/native/libnio/fs/UnixNativeDispatcher.c
@@ -80,5 +80,5 @@
 // not relying on included headers.
 
-#ifndef __GLIBC__
+#if !defined(__GLIBC__) && !defined(__ANDROID__)
 // Alpine doesn't know these types, define them
 typedef unsigned int       __uint32_t;
