// skip some checks as AOSP does
diff --git a/src/java.base/unix/native/libnet/net_util_md.c b/src/java.base/unix/native/libnet/net_util_md.c
index 4ec11a136..01b85db4d 100644
--- a/src/java.base/unix/native/libnet/net_util_md.c
+++ b/src/java.base/unix/native/libnet/net_util_md.c
@@ -129,6 +129,7 @@ jint  IPv6_supported()
     SOCKETADDRESS sa;
     socklen_t sa_len = sizeof(SOCKETADDRESS);
 
+#ifndef __ANDROID__  // ANDROID: skip check, see libcore commit ae218d9b
     fd = socket(AF_INET6, SOCK_STREAM, 0) ;
     if (fd < 0) {
         /*
@@ -172,6 +173,7 @@ jint  IPv6_supported()
         }
     }
 #endif
+#endif  // !defined __ANDROID__
 
     /*
      *  OK we may have the stack available in the kernel,
